### PR TITLE
CORDA-2190: Improve notary service flow exception handling

### PR DESCRIPTION
**Pull request by adagys**
- Refactored notary flows to reduce validation code duplication
- Improved notarisation error handling to provide more helpful responses to the client

### CODE REVIEW

File: core/src/main/kotlin/net/corda/core/internal/notary/NotaryUtils.kt [hunk 1]
```diff
@@ -2,42 +2,30 @@ package net.corda.core.internal.notary
...
+            "Expected a signature by ${intendedSigner.owningKey}, but received by ${signature.by}}"
         }
+
+        // TODO: if requestSignature was generated over an old version of NotarisationRequest, we need to be able to
+        // reserialize it in that version to get the exact same bytes. Modify the serialization logic once that's
+        // available.
+        val expectedSignedBytes = this.serialize().bytes
+        signature.verify(expectedSignedBytes)
+    } catch (e: Exception) {
+        val error = NotaryError.RequestSignatureInvalid(e)
```
Comment (rick-r3) [inline core/src/main/kotlin/net/corda/core/internal/notary/NotaryUtils.kt]: We will mask the real cause here still.  Surely we want to log `e` ?

File: core/src/main/kotlin/net/corda/core/internal/notary/NotaryServiceFlow.kt [hunk 2]
```diff
@@ -23,61 +23,84 @@ abstract class NotaryServiceFlow(val otherSideSession: FlowSession, val service:
...
-            )
-            signTransactionAndSendResponse(txId)
+                    tx.timeWindow,
+                    tx.references)
+
         } catch (e: NotaryInternalException) {
-            throw NotaryException(e.error, txId)
+            // Any exception that's not a NotaryInternalException is assumed to be an unexpected internal error
+            // that is not relayed back to the client.
+            throw NotaryException(e.error, transactionId)
```
Comment (rick-r3) [inline core/src/main/kotlin/net/corda/core/internal/notary/NotaryServiceFlow.kt]: Are you sure it's not just easier to log the cause here?  i.e. use `when` on `e.error` to pick out those with a `cause` and log it.

Ordinarily we'd perhaps modify `NotaryError` so we could do that in an OO way, but wire stability and all that...

Comment (adagys) [reply core/src/main/kotlin/net/corda/core/internal/notary/NotaryServiceFlow.kt]: Yeah actually that would be better

**Comment by adagys** [comment]
Comments addressed. This is what gets logged now when a transaction verification failure occurs:
```
[ERROR] 09:50:04,789 [Mock network] corda.flow.logError - Error notarising transaction DC15D8D4FB6FAB436B11E9AAF66EE8E858E756B7D087E5CAB7DBB55D1A832759 [errorCode=kphuh9] {fiber-id=10000002, flow-id=b13a0c54-2e8f-4b24-a4ff-10ba37e3f1f3, invocation_id=03c40f0d-0cac-4455-826f-472e356661eb, invocation_timestamp=2018-11-09T09:50:04.619Z, origin=O=Alice Corp, L=Madrid, C=ES, session_id=03c40f0d-0cac-4455-826f-472e356661eb, session_timestamp=2018-11-09T09:50:04.619Z, thread-id=135}
 net.corda.core.transactions.SignedTransaction$SignaturesMissingException: Missing signatures for [notary] on transaction B6CBC5 for net.i2p.crypto.eddsa.EdDSAPublicKey@5fce14ab
	at net.corda.core.transactions.TransactionWithSignatures.verifySignaturesExcept(TransactionWithSignatures.kt:75) ~[classes/:?]
	at net.corda.core.transactions.TransactionWithSignatures.verifyRequiredSignatures(TransactionWithSignatures.kt:38) ~[classes/:?]
	at net.corda.core.transactions.SignedTransaction.toLedgerTransaction(SignedTransaction.kt:151) ~[classes/:?]
	at net.corda.core.transactions.SignedTransaction.verifyRegularTransaction(SignedTransaction.kt:200) ~[classes/:?]
	at net.corda.core.transactions.SignedTransaction.verify(SignedTransaction.kt:175) ~[classes/:?]
	at net.corda.core.transactions.SignedTransaction.verify$default(SignedTransaction.kt:171) ~[classes/:?]
	at net.corda.core.internal.ResolveTransactionsFlow.call(ResolveTransactionsFlow.kt:95) ~[classes/:?]
	at net.corda.core.internal.ResolveTransactionsFlow.call(ResolveTransactionsFlow.kt:27) ~[classes/:?]
	at net.corda.node.services.statemachine.FlowStateMachineImpl.subFlow(FlowStateMachineImpl.kt:275) ~[classes/:?]
	at net.corda.core.flows.FlowLogic.subFlow(FlowLogic.kt:309) ~[classes/:?]
	at net.corda.node.services.transactions.ValidatingNotaryFlow.resolveAndContractVerify(ValidatingNotaryFlow.kt:46) ~[classes/:?]
	at net.corda.node.services.transactions.ValidatingNotaryFlow.verifyTransaction(ValidatingNotaryFlow.kt:37) ~[classes/:?]
	at net.corda.core.internal.notary.NotaryServiceFlow.call(NotaryServiceFlow.kt:40) [classes/:?]
	at net.corda.core.internal.notary.NotaryServiceFlow.call(NotaryServiceFlow.kt:20) [classes/:?]
	at net.corda.node.services.statemachine.FlowStateMachineImpl.run(FlowStateMachineImpl.kt:219) [classes/:?]
	at net.corda.node.services.statemachine.FlowStateMachineImpl.run(FlowStateMachineImpl.kt:42) [classes/:?]
	at co.paralleluniverse.fibers.Fiber.run1(Fiber.java:1092) [quasar-core-0.7.10-jdk8.jar:0.7.10]
	at co.paralleluniverse.fibers.Fiber.exec(Fiber.java:788) [quasar-core-0.7.10-jdk8.jar:0.7.10]
	at co.paralleluniverse.fibers.RunnableFiberTask.doExec(RunnableFiberTask.java:100) [quasar-core-0.7.10-jdk8.jar:0.7.10]
	at co.paralleluniverse.fibers.RunnableFiberTask.run(RunnableFiberTask.java:91) [quasar-core-0.7.10-jdk8.jar:0.7.10]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [?:1.8.0_171]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:1.8.0_171]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180) [?:1.8.0_171]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293) [?:1.8.0_171]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_171]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_171]
	at net.corda.node.utilities.AffinityExecutor$ServiceAffinityExecutor$1$thread$1.run(AffinityExecutor.kt:63) [classes/:?]
[INFO] 09:50:04,795 [Mock network] corda.flow.run - Flow threw exception... sending it to flow hospital {fiber-id=10000002, flow-id=b13a0c54-2e8f-4b24-a4ff-10ba37e3f1f3, invocation_id=03c40f0d-0cac-4455-826f-472e356661eb, invocation_timestamp=2018-11-09T09:50:04.619Z, origin=O=Alice Corp, L=Madrid, C=ES, session_id=03c40f0d-0cac-4455-826f-472e356661eb, session_timestamp=2018-11-09T09:50:04.619Z, thread-id=135}
 net.corda.core.flows.NotaryException: Unable to notarise transaction DC15D8D4FB6FAB436B11E9AAF66EE8E858E756B7D087E5CAB7DBB55D1A832759 : net.corda.core.transactions.SignedTransaction$SignaturesMissingException: Missing signatures for [notary] on transaction B6CBC5 for net.i2p.crypto.eddsa.EdDSAPublicKey@5fce14ab
	at net.corda.core.internal.notary.NotaryServiceFlow.call(NotaryServiceFlow.kt:54) ~[classes/:?]
	at net.corda.core.internal.notary.NotaryServiceFlow.call(NotaryServiceFlow.kt:20) ~[classes/:?]
	at net.corda.node.services.statemachine.FlowStateMachineImpl.run(FlowStateMachineImpl.kt:219) [classes/:?]
	at net.corda.node.services.statemachine.FlowStateMachineImpl.run(FlowStateMachineImpl.kt:42) [classes/:?]
	at co.paralleluniverse.fibers.Fiber.run1(Fiber.java:1092) [quasar-core-0.7.10-jdk8.jar:0.7.10]
	at co.paralleluniverse.fibers.Fiber.exec(Fiber.java:788) [quasar-core-0.7.10-jdk8.jar:0.7.10]
	at co.paralleluniverse.fibers.RunnableFiberTask.doExec(RunnableFiberTask.java:100) [quasar-core-0.7.10-jdk8.jar:0.7.10]
	at co.paralleluniverse.fibers.RunnableFiberTask.run(RunnableFiberTask.java:91) [quasar-core-0.7.10-jdk8.jar:0.7.10]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [?:1.8.0_171]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:1.8.0_171]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180) [?:1.8.0_171]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293) [?:1.8.0_171]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_171]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_171]
	at net.corda.node.utilities.AffinityExecutor$ServiceAffinityExecutor$1$thread$1.run(AffinityExecutor.kt:63) [classes/:?]
```

File: core/src/main/kotlin/net/corda/core/internal/notary/NotaryServiceFlow.kt [hunk 3]
```diff
@@ -23,61 +23,85 @@ abstract class NotaryServiceFlow(val otherSideSession: FlowSession, val service:
...
-            )
+    private fun validateRequest(requestPayload: NotarisationPayload): TransactionParts {
+        try {
+            val transaction = extractParts(requestPayload)
+            transactionId = transaction.id
+            checkNotary(transaction.notary)
+            checkInputs(transaction.inputs + transaction.references)
+            return transaction
+        } catch (e: Exception) {
+            logger.error("Transaction invalid", e)
```
Comment (rick-r3) [inline core/src/main/kotlin/net/corda/core/internal/notary/NotaryServiceFlow.kt]: We can remove this can't we?

Comment (adagys) [reply core/src/main/kotlin/net/corda/core/internal/notary/NotaryServiceFlow.kt]: Yes, missed that one

**Comment by gendal** [comment]
I was just browsing through a pile of PRs from the last week or so and got thinking... is the effect of this PR entirely internal? Or should it be mentioned in release-notes? changelog?  anything needing to be updated in docs?

**Comment by adagys** [comment]
Apart from adding more logging, it only affects notary implementations, for which we explicity don't guarantee stability. There is a page in the docs with an example notary that is updated.

**Pull request merged**